File: core/src/main/java/pcb/uwu/core/CompositeUnit.java
package pcb.uwu.core;

import pcb.uwu.core.UnitCounter.UnitCount;

import java.util.Objects;
import java.util.function.Function;

import static pcb.uwu.core.BigDecimalAmount.ONE;
import static pcb.uwu.core.BigDecimalAmount.ZERO;

public class CompositeUnit implements Unit {

	// region private fields

	private final UnitCounter unitCounter;

	// endregion

	// region constructors

	CompositeUnit() {
		this(new UnitCounter());
	}

	public CompositeUnit(UnitCounter unitCounter)
	{
		this.unitCounter = unitCounter;
	}

	// endregion

	// region public methods

	public CompositeUnit major(BaseUnit unit) {
		return new CompositeUnit(unitCounter.major(unit));
	}

	public CompositeUnit minor(BaseUnit unit) {
		return new CompositeUnit(unitCounter.minor(unit));
	}

	public UnitCounter getUnitCounter() {
		return unitCounter;
	}

	// endregion

	// region implement Unit

	/**
	 * {@inheritDoc}
	 */
	@Override
	public String getSymbol() {
		return unitCounter.asString(Unit::getSymbol, Unit::getSymbol, "/");
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public String getSingularName() {
		return unitCounter.asString(Unit::getSingularName, Unit::getSingularName, " per ");
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public String getPluralName() {
		return unitCounter.asString(Unit::getPluralName, Unit::getSingularName, " per ");
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Function<BigDecimalAmount, BigDecimalAmount> getTranslationToCanonical() {
		Function<BigDecimalAmount, BigDecimalAmount> result = Function.identity();

		for (UnitCount unitCount : unitCounter.getBaseUnits()) {
			BaseUnit unit = unitCount.getUnit();
			int count = unitCount.getCount();

			if (count > 0) {
				for (int i = 0; i < count; i++) {
					result = result.andThen(unit.getTranslationToCanonical());
				}
			} else if (count < 0) {
				for (int i = 0; i > count; i--) {
					result = result.andThen(unit.getTranslationFromCanonical());
				}
			}
		}

		return result;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public Function<BigDecimalAmount, BigDecimalAmount> getTranslationFromCanonical() {
		Function<BigDecimalAmount, BigDecimalAmount> result = Function.identity();

		for (UnitCount unitCount : unitCounter.getBaseUnits()) {
			BaseUnit unit = unitCount.getUnit();
			int count = unitCount.getCount();

			if (count > 0) {
				for (int i = 0; i < count; i++) {
					result = result.andThen(unit.getTranslationFromCanonical());
				}
			} else if (count < 0) {
				for (int i = 0; i > count; i--) {
					result = result.andThen(unit.getTranslationToCanonical());
				}
			}
		}

		return result;
	}

	/**
	 * {@inheritDoc}
	 */
	@Override
	public boolean isScalar() {
		return unitCounter.isEmpty();
	}

	@Override
	public Class<? extends BaseUnit> getBaseUnitType() {
		return null;
	}

	// endregion

	// region override Object

	@Override
	public boolean equals(Object obj) {
		if (obj instanceof Unit) {
			Unit that = (Unit) obj;

			return Objects.equals(this.unitCounter, that.getUnitCounter())
					&& Objects.equals(this.getTranslationFromCanonical().apply(ONE), that.getTranslationFromCanonical().apply(ONE))
					&& Objects.equals(this.getTranslationToCanonical().apply(ONE), that.getTranslationToCanonical().apply(ONE));
		}

		return false;
	}

	@Override
	public int hashCode() {
		return unitCounter.hashCode();
	}

	@Override
	public String toString() {
		return getSymbol();
	}

	// endregion
}

File: core/src/main/java/pcb/uwu/core/CompositeUnit.kt
package pcb.uwu.core

import java.util.function.Function

open class CompositeUnit(final override val unitCounter: UnitCounter = UnitCounter())
    : Unit
{
    // region public methods

    fun major(unit: BaseUnit) =
        CompositeUnit(unitCounter.major(unit))

    fun minor(unit: BaseUnit) =
        CompositeUnit(unitCounter.minor(unit))

    // endregion

    // region Unit

    override val symbol = unitCounter.asString(Unit::symbol, Unit::symbol)

    override val singularName = unitCounter.asString(Unit::singularName, Unit::singularName)

    override val pluralName = unitCounter.asString(Unit::pluralName, Unit::singularName)

    override val translationToCanonical: Function<Amount, Amount>
        get()
        {
            var result = Function.identity<Amount>()
            for (unitCount in unitCounter.baseUnits)
            {
                val unit = unitCount.unit
                val count = unitCount.count
                if (count > 0)
                {
                    for (i in 0 until count)
                    {
                        result = result.andThen(unit.translationToCanonical)
                    }
                }
                else if (count < 0)
                {
                    for (i in 0 downTo count + 1)
                    {
                        result = result.andThen(unit.translationFromCanonical)
                    }
                }
            }
            return result
        }

    override val translationFromCanonical: Function<Amount, Amount>
        get()
        {
            var result = Function.identity<Amount>()
            for (unitCount in unitCounter.baseUnits)
            {
                val unit = unitCount.unit
                val count = unitCount.count
                if (count > 0)
                {
                    for (i in 0 until count)
                    {
                        result = result.andThen(unit.translationFromCanonical)
                    }
                }
                else if (count < 0)
                {
                    for (i in 0 downTo count + 1)
                    {
                        result = result.andThen(unit.translationToCanonical)
                    }
                }
            }
            return result
        }

    override val isScalar = unitCounter.isEmpty

    override val baseUnitType: Class<out BaseUnit>
        get() = throw IllegalStateException()

    // endregion

    // region Object

    @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
    override fun equals(that: Any?): Boolean
    {
        if (this === that) return true
        if (that !is Unit) return false
        return unitCounter == that.unitCounter &&
                translationFromCanonical.apply(Amount.ONE) == that.translationFromCanonical.apply(
                Amount.ONE) &&
                translationToCanonical.apply(Amount.ONE) == that.translationToCanonical.apply(Amount.ONE)
    }

    override fun hashCode() = unitCounter.hashCode()

    override fun toString() = symbol

    // endregion
}